### ISSUE TITLE
Question: Is it possible to dynamically produce nodes where the number of the nodes are determined by some variable?

### ISSUE THREAD

**Issue by hrcHarry** [L3MON4D3/LuaSnip#914]
Hi

The _idea_ will be like the following (invalid) setting
<pre><code>local function produce_nodes(args, snip)
  local vars_node_start_idx = 0
  local var_num = snip.captures[1]
    
  local nodes_table = {}
    
  if ((var_num ~= nil) and (var_num >= 1)) then
    for j = 1, var_num do
      var_type_start_idx = vars_node_start_idx + 1
      var_name_start_idx = vars_node_start_idx + 2
      table.insert(nodes_table, sn(nil, {i(var_type_start_idx, "var_type"),
                                         t(" "),
                                         i(var_name_start_idx, "var_name"),
                                         f(function() 
                                              if j == var_num then
                                                return t({""})
                                              else
                                                return t({",", ""})
                                           end)}))
      vars_node_start_idx = vars_node_start_idx + 2    
    end
  end  
  return sn(nil, nodes_table)
end

s({trig="trig(%d)v", regTrig=true, snippetType="autosnippet"},
  fmta( [[
           < return_type > < func_name >(< vars >)
           {
              < func_body >
           }
        ]],
        {
          return_type = i(1, "return_type"),
          func_name = i(2, "func_name"),
          vars = d(3, produce_nodes),
          func_body = i(4, "// func_body")
        } 
  )
)</code></pre>
In summary, `(%d)` in trigger _capture_ the number of the variable we would like to define. 
For example, if I type `trig3v`, I expect that it will result in 
<pre><code> return_type func_name(var_type var_name,
                       var_type var_name,
                       var_type var_name)
{
  // function body
}</code></pre>
After I read the DOC, I have no idea to implement it, and I'm not sure if it can be carried out.
Could someone have any idea? Thanks.

**Comment by hrcHarry** [comment]
Hi. Thanks for replying. 
Taking your suggestion (in the `dynamicNode`, passing `i()`s to `sn()` by `vim.list_extend()` and returning that `sn()`), I got the result triggered by `trig3v`
<pre><code>return_type func_name(var_type var_name,
var_type var_name,
var_type var_name)
{
  // function body
}</code></pre>
Every node can be edited. That is awesome! 

But now I have a new situation: the `insertNode` do not support indentation while my expectation is
<pre><code>return_type func_name(var_type var_name,
                      var_type var_name,
                      var_type var_name)
{
  // function body
}</code></pre>

Now I got the situation to fix this. 

At first, I thought `indentSnippetNode` may help but I realized that if I "insert" the `return_type` and `func_name`, the following `var_type`s and `var_name`s will not comply with indentation. I mean, for example, if I "insert" the `return_type` and `func_name` by `MyType thisIsATestFunction`, it may result in
<pre><code>MyType thisIsATestFunction(var_type var_name,
                      var_type var_name,
                      var_type var_name)
{
  // function body
}</code></pre>
However, my expectation is 
<pre><code>MyType thisIsATestFunction(var_type var_name,
                           var_type var_name,
                           var_type var_name)
{
  // function body
}</code></pre>
I think it may be fixed by the function in `dynamicNode`. 
I come up with the first idea that to get the lengths of strings, in this case, `#"MyType"` and `#"thisIsATestFunction"` and using them in the function in `dynamicNode`. 
Unfortunately, it seems that [Snippet-Data](https://github.com/L3MON4D3/LuaSnip/blob/master/DOC.md#data) does not offer the related information.

Is there any way to get this? Or some other alternative way to accomplish the indentation while "updating" the node? Thanks.

**Comment by L3MON4D3** [comment]
Oh cool, haven't come across this yet :D
I think you can have the dynamicNode depend on the first two nodes, and then get their lengths in the dynamicNode-function

**Comment by hrcHarry** [comment]
Hi.
I've tried making `dynamicNode` depending on the first two nodes, and I got the error
![](https://upload.cc/i1/2023/06/10/BbeGqV.png)

My setting is shown as follows
<pre><code>-- cases
local indentation_case = {     
  [0] = function()
          return {t("")}
        end,
  [1] = function()
          return {i(1, "var_type"), t(" "), i(2, "var_name")}
        end,
}

-- default
local function indentation_default(args, snip)     
  local vars_node_jump_idx = 4
  local var_num = tonumber(snip.captures[1])
  local length_indentation = #(args[1][1]) + #(args[2][1]) + 1
  local indentation_spaces = ""
  local loop_start_idx_when_var_num_greater_than_2 = 3
  local nodes_table = {}

  for j = 1, length_indentation do
    indentation_spaces = indentation_spaces .. " "
  end

  nodes_table = {i(1, "var_type"), t(" "), i(2, "var_name"), t({"", indentation_spaces}), i(3, "var_type"), t(" "), i(4, "var_name")}

  if (var_num >= loop_start_idx_when_var_num_greater_than_2) then

    for j = loop_start_idx_when_var_num_greater_than_2, (var_num-1) do
      var_type_jump_idx = vars_node_jump_idx + 1
      var_name_jump_idx = vars_node_jump_idx + 2

      vim.list_extend(nodes_table, {t({"", indentation_spaces}), i(var_type_jump_idx, "var_type"), t(" "), i(var_name_jump_idx, "var_name"), t({",", ""})})

      vars_node_jump_idx = vars_node_jump_idx + 2
    end

    var_type_jump_idx = vars_node_jump_idx + 1
    var_name_jump_idx = vars_node_jump_idx + 2
    vim.list_extend(nodes_table, {t({"", indentation_spaces}), i(var_type_jump_idx, "var_type"), t(" "), i(var_name_jump_idx, "var_name")})

  end

  return nodes_table
end

-- switch cases
local function switch_indentation_case(case_table, table_idx, default_func, args, snip)
  if (case_table[table_idx] ~= nil) then
    case_table[table_idx]()
  else
    default_func(args, snip)
  end
end

-- function in dynamicNode
local function produce_nodes_4(args, snip)
  local vars_node_start_idx = 0
  local var_num = tonumber(snip.captures[1])

  local nodes_table_in_sn = {}

  if ((var_num ~= nil) ) then
    nodes_table_in_sn = switch_indentation_case(indentation_case, var_num, indentation_default, args, snip)
  else
    nodes_table_in_sn = {t("")}
  end

  return sn(nil, nodes_table_in_sn)
end</code></pre>

And the snippet is set as follows
<pre><code>  s({trig="trig(%d)v", regTrig=true, snippetType="autosnippet"},
    fmta( [[
             < return_type > < func_name >(< vars >)
             {
                < func_body >
             }
          ]],
          {
            return_type = i(1, "return_type"),
            func_name = i(2, "func_name"),
            vars = d(3, produce_nodes_4, {1, 2}),
            func_body = i(4, "// func_body")
          }
    )
  )</code></pre>

I'm not sure where is the mistake. Do I misunderstand some point in `dynamicNode`?

**Comment by L3MON4D3** [comment]
I think it's that `nodes_table_in_sn` is assigned `nil` from `switch_indentation_case`

**Comment by hrcHarry** [comment]
Ohhhh, my bad. I got the sloppy mistake in `switch_indentation_case`. 😥😓
Appreciate your patience and generosity 😅

I have accomplished my desired result and I think it can be closed.
Thank you again.

**Comment by L3MON4D3** [comment]
Happy to help! :D
